File: examples/awfy/go/json.go
package main

// Fixed input exercising objects, arrays, strings, integers, bools, null.
// The identical literal is embedded in all three languages.
const jsonInput = `{"widget":{"debug":"on","window":{"title":"Sample","width":500,"height":300},"items":[1,2,3,4,5],"enabled":true,"data":null}}`

type jsonKind int

const (
	jNull jsonKind = iota
	jBool
	jNum
	jStr
	jArr
	jObj
)

type jsonEntry struct {
	key   string
	value *jsonValue
}

type jsonValue struct {
	kind    jsonKind
	num     int
	str     string
	arr     []*jsonValue
	obj     []jsonEntry
}

type jsonParser struct {
	s string
	i int
}

func (p *jsonParser) skipWs() {
	for p.i < len(p.s) {
		c := p.s[p.i]
		if c == ' ' || c == '\n' || c == '\t' || c == '\r' {
			p.i++
		} else {
			break
		}
	}
}

func (p *jsonParser) parseStr() string {
	p.i++ // opening quote
	start := p.i
	for p.s[p.i] != '"' {
		p.i++
	}
	out := p.s[start:p.i]
	p.i++ // closing quote
	return out
}

func (p *jsonParser) parseNum() int {
	neg := false
	if p.s[p.i] == '-' {
		neg = true
		p.i++
	}
	n := 0
	for p.i < len(p.s) && p.s[p.i] >= '0' && p.s[p.i] <= '9' {
		n = n*10 + int(p.s[p.i]-'0')
		p.i++
	}
	if neg {
		return -n
	}
	return n
}

func (p *jsonParser) parseValue() *jsonValue {
	p.skipWs()
	c := p.s[p.i]
	switch {
	case c == '{':
		return p.parseObj()
	case c == '[':
		return p.parseArr()
	case c == '"':
		return &jsonValue{kind: jStr, str: p.parseStr()}
	case c == 't':
		p.i += 4
		return &jsonValue{kind: jBool}
	case c == 'f':
		p.i += 5
		return &jsonValue{kind: jBool}
	case c == 'n':
		p.i += 4
		return &jsonValue{kind: jNull}
	default:
		return &jsonValue{kind: jNum, num: p.parseNum()}
	}
}

func (p *jsonParser) parseArr() *jsonValue {
	p.i++ // [
	items := []*jsonValue{}
	p.skipWs()
	if p.s[p.i] == ']' {
		p.i++
		return &jsonValue{kind: jArr, arr: items}
	}
	for {
		items = append(items, p.parseValue())
		p.skipWs()
		if p.s[p.i] == ',' {
			p.i++
			continue
		}
		p.i++ // ]
		break
	}
	return &jsonValue{kind: jArr, arr: items}
}

func (p *jsonParser) parseObj() *jsonValue {
	p.i++ // {
	entries := []jsonEntry{}
	p.skipWs()
	if p.s[p.i] == '}' {
		p.i++
		return &jsonValue{kind: jObj, obj: entries}
	}
	for {
		p.skipWs()
		key := p.parseStr()
		p.skipWs()
		p.i++ // :
		v := p.parseValue()
		entries = append(entries, jsonEntry{key: key, value: v})
		p.skipWs()
		if p.s[p.i] == ',' {
			p.i++
			continue
		}
		p.i++ // }
		break
	}
	return &jsonValue{kind: jObj, obj: entries}
}

func jsonFold(n *jsonValue, sum, count *int) {
	*count++
	switch n.kind {
	case jNum:
		*sum += n.num
	case jArr:
		for _, x := range n.arr {
			jsonFold(x, sum, count)
		}
	case jObj:
		for _, e := range n.obj {
			jsonFold(e.value, sum, count)
		}
	}
}

func jsonRun(size int) int {
	checksum := 0
	for r := 0; r < size; r++ {
		p := &jsonParser{s: jsonInput, i: 0}
		tree := p.parseValue()
		sum, count := 0, 0
		jsonFold(tree, &sum, &count)
		checksum = sum*31 + count
	}
	return checksum
}

var jsonBench = Bench{Name: "json", Warmup: 20, Iters: 100, Size: 1000, Expected: 25280, Run: jsonRun}

File: examples/performance/awfy/node/json.mjs
export const warmup = 20, iters = 100, size = 1000, expected = 25280;

// Fixed input exercising objects, arrays, strings, integers, bools, null.
// The identical literal is embedded in all three languages.
export const INPUT =
  '{"widget":{"debug":"on","window":{"title":"Sample","width":500,"height":300},"items":[1,2,3,4,5],"enabled":true,"data":null}}';

// A tiny hand-written recursive-descent parser (no JSON.parse; the point is
// comparable parser codegen). Values: {null,bool,num,str,arr,obj}.
function parse(s) {
  let i = 0;
  const skipWs = () => { while (i < s.length && (s[i] === ' ' || s[i] === '\n' || s[i] === '\t' || s[i] === '\r')) i++; };
  function value() {
    skipWs();
    const c = s[i];
    if (c === '{') return obj();
    if (c === '[') return arr();
    if (c === '"') return { t: "str", v: str() };
    if (c === 't') { i += 4; return { t: "bool", v: true }; }
    if (c === 'f') { i += 5; return { t: "bool", v: false }; }
    if (c === 'n') { i += 4; return { t: "null" }; }
    return { t: "num", v: num() };
  }
  function str() {
    i++; // opening quote
    let start = i;
    while (s[i] !== '"') i++;
    const out = s.slice(start, i);
    i++; // closing quote
    return out;
  }
  function num() {
    let start = i;
    if (s[i] === '-') i++;
    while (i < s.length && s[i] >= '0' && s[i] <= '9') i++;
    return parseInt(s.slice(start, i), 10);
  }
  function arr() {
    i++; // [
    const items = [];
    skipWs();
    if (s[i] === ']') { i++; return { t: "arr", v: items }; }
    while (true) {
      items.push(value());
      skipWs();
      if (s[i] === ',') { i++; continue; }
      i++; // ]
      break;
    }
    return { t: "arr", v: items };
  }
  function obj() {
    i++; // {
    const entries = [];
    skipWs();
    if (s[i] === '}') { i++; return { t: "obj", v: entries }; }
    while (true) {
      skipWs();
      const key = str();
      skipWs();
      i++; // :
      const v = value();
      entries.push({ key, value: v });
      skipWs();
      if (s[i] === ',') { i++; continue; }
      i++; // }
      break;
    }
    return { t: "obj", v: entries };
  }
  return value();
}

// Fold: count every parsed node and sum all integers.
function fold(node, acc) {
  acc.count++;
  switch (node.t) {
    case "num": acc.sum += node.v; break;
    case "arr": for (const x of node.v) fold(x, acc); break;
    case "obj": for (const e of node.v) fold(e.value, acc); break;
  }
  return acc;
}

export function run(size) {
  let checksum = 0;
  for (let r = 0; r < size; r++) {
    const tree = parse(INPUT);
    const acc = fold(tree, { sum: 0, count: 0 });
    checksum = acc.sum * 31 + acc.count;
  }
  return checksum;
}
